source/priority_queue.c

```diff
@@ -0,0 +1,135 @@
+/*
+* Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/priority_queue.h>
+#include <string.h>
+
+static void make_heap_from_top(struct aws_priority_queue *queue, int index) {
+    int left = (index << 1) + 1;
+    int right = left + 1;
+    int swap_candidate = index;
+    int len = aws_array_list_length(&queue->container);
+
+
+    void *item, *swap_item;
+    aws_array_list_get_at_ptr(&queue->container, &item, left);
+    aws_array_list_get_at_ptr(&queue->container, &swap_item, swap_candidate);
+
+    if (left < len && queue->compare(item, swap_item)) {
```

Comment:
What are the semantics of compare? Are we swapping whenever the items are nonequal here?